src/clients/http.rs

```diff
@@ -0,0 +1,139 @@
+            health_url,
+            client,
+        }
+    }
+
+    pub fn base_url(&self) -> &Url {
+        &self.base_url
+    }
+
+    /// This is sectioned off to allow for testing.
+    pub(super) async fn http_response_to_health_check_result(
+        res: Result<Response, reqwest::Error>,
+    ) -> HealthCheckResult {
+        match res {
+            Ok(response) => {
+                if response.status() == StatusCode::OK {
+                    if let Ok(body) = response.json::<OptionalHealthCheckResponseBody>().await {
+                        // If the service provided a body, we only anticipate a minimal health status and optional reason.
+                        HealthCheckResult {
+                            status: body.status.clone(),
+                            code: StatusCode::OK,
+                            reason: match body.status {
+                                HealthStatus::Healthy => None,
+                                _ => body.reason,
+                            },
+                        }
+                    } else {
+                        // If the service did not provide a body, we assume it is healthy.
+                        HealthCheckResult {
+                            status: HealthStatus::Healthy,
+                            code: StatusCode::OK,
+                            reason: None,
+                        }
+                    }
+                } else {
+                    HealthCheckResult {
+                        // The most we can presume is that 5xx errors are likely indicating service issues, implying the service is unhealthy.
+                        // and that 4xx errors are more likely indicating health check failures, i.e. due to configuration/implementation issues.
+                        // Regardless we can't be certain, so the reason is also provided.
+                        // TODO: We will likely circle back to re-evaluate this logic in the future
+                        // when we know more about how the client health results will be used.
+                        status: if response.status().as_u16() >= 500
+                            && response.status().as_u16() < 600
+                        {
+                            HealthStatus::Unhealthy
+                        } else if response.status().as_u16() >= 400
+                            && response.status().as_u16() < 500
+                        {
+                            HealthStatus::Unknown
+                        } else {
+                            error!(
+                                "unexpected http health check status code: {}",
+                                response.status()
+                            );
+                            HealthStatus::Unknown
+                        },
+                        code: response.status(),
+                        reason: response.status().canonical_reason().map(|v| v.to_string()),
+                    }
+                }
```

Comment:
Should we move this to a separate (probably inline) function to reduce nesting, or is this common in Rust?

Reply:
I thought about trying to further simplify/reduce this logic, but it wasn't added as part of this PR, so I tried to keep it as close as possible. I did change/simplify how we build the `reason` values though.  I don't think the nesting is an issue here as its only 2 levels of `if/else`.